pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go

```diff
@@ -141,58 +142,26 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 			fmt.Sprintf("%s/%s", pr.Namespace, pr.Spec.PipelineRef.Name))
 		return nil
 	}
-	if err := c.createPipelineRunTaskRuns(p, pr); err != nil {
-		return err
-	}
-	// TODO fetch the taskruns status for this pipeline run.
-
-	// TODO check status of tasks and update status of PipelineRuns
-
-	return nil
-}
-
-// createPipelineRunTaskRuns goes through all the `Task` in the `PipelineSpec`
-// and creates a `TaskRun` for the first `Task` for which no corresponding
-// `TaskRun` exist.
-func (c *Reconciler) createPipelineRunTaskRuns(p *v1alpha1.Pipeline, pr *v1alpha1.PipelineRun) error {
-	taskRefMap, err := c.getTaskMap(p)
+	pipelineTasks, err := pipeline.GetTasks(c.taskLister, p)
 	if err != nil {
-		return errors.NewBadRequest(fmt.Sprintf("pipeline %q is not valid due to %v", p.Name, err))
+		return fmt.Errorf("error getting map of created TaskRuns for Pipeline %s: %s", p.Name, err)
 	}
-	for _, pt := range p.Spec.Tasks {
-		trName := getTaskRunName(pr.Name, &pt)
-		_, listErr := c.taskRunLister.TaskRuns(p.Namespace).Get(trName)
-		if errors.IsNotFound(listErr) && c.canTaskRun(&pt) {
-			if _, err := c.createTaskRun(taskRefMap[pt.Name], trName, pr); err != nil {
-				return err
-			}
-		} else if err != nil {
-			return err
-		}
+	pipelineTaskName, trName, err := getNextPipelineRunTaskRun(c.taskRunLister, p, pr.Name)
+	if err != nil {
+		return fmt.Errorf("error getting next TaskRun to create for PipelineRun %s: %s", pr.Name, err)
 	}
-	return nil
-}
-
-// getTaskMap returns a map of task in pipeline spec to `Task` mentioned in taskRef.
-func (c *Reconciler) getTaskMap(p *v1alpha1.Pipeline) (map[string]*v1alpha1.Task, error) {
-	tMap := make(map[string]*v1alpha1.Task, 0)
-	for _, pt := range p.Spec.Tasks {
-		tObj, err := c.taskLister.Tasks(p.Namespace).Get(pt.TaskRef.Name)
+	if pipelineTaskName != "" {
+		_, err = c.createTaskRun(pipelineTasks[pipelineTaskName], trName, pr)
 		if err != nil {
-			c.Logger.Errorf("failed to get tasks for Pipeline %q: Error getting task %q : %s",
-				fmt.Sprintf("%s/%s", p.Namespace, p.Name),
-				fmt.Sprintf("%s/%s", p.Namespace, pt.TaskRef.Name), err)
-			return nil, err
+			return fmt.Errorf("error creating TaskRun called %s for PipelineTask %s from PipelineRun %s: %s", trName, pipelineTaskName, pr.Name, err)
```

Comment:
hm okay - seems weird that we want to controller to forget tho - what if the Task got recreated or something?

ill do this for now tho to be consistent with how it was working before

Reply:
hm it looks like the previous behaviour tho was to return an error in these cases:

1. if getTaskMap returned an error (i.e. the pipeline is invalid):
![image](https://user-images.githubusercontent.com/432502/46702748-2726dc00-cbd9-11e8-9a69-f9ac29519eb8.png)
![image](https://user-images.githubusercontent.com/432502/46702752-2f7f1700-cbd9-11e8-84b5-b3fdd39c2b6e.png)

2. If createTask failed:
![image](https://user-images.githubusercontent.com/432502/46702813-6fde9500-cbd9-11e8-9ce1-4b9c4aaa14d6.png)

I _think_ the new code is behaving in the same way, it might be confusing in this diff b/c of the way the code is overlaid - might help to look at the tests and see if errors are happening in the cases you expect? (or maybe we need to add some more cases)

Reply:
I think Returning an error is the right thing to do. If the Task doesn’t exist we can’t proceed. Same for if we can’t create the TaskRun